### PR TITLE
Avoid printing comments on static fields twice.

### LINKED ISSUE #45696: Synthetic comment duplicated on class with decorators
# Bug Report

### 🔎 Search Terms
`setSyntheticLeadingComments`, `updateClassDeclaration`, `transformer`

### 🕗 Version & Regression Information

My project has a custom transformer that declares new `static` properties on classes and attaches comments to them using the `setSyntheticLeadingComments` function. In TypeScript version 4.4 the comments are duplicated if the class has at least one decorator. If the class doesn't have decorators, the comments are emitted only once.

- This changed between versions 4.3.5 and 4.4.2

### ⏯ Playground Link
Since this is a problem with the compiler API, I can't link to the playground. Instead I have isolated the issue in a small repository. 

### **[Link to the reproduction repository](https://github.com/crisbeto/ts-4.4-synthetic-comment)**

To reproduce the issue, do the follwoing:
1. `git clone https://github.com/crisbeto/ts-4.4-synthetic-comment.git`
2. `npm install` inside the repo.
3. `npm run transform`.
4. Observe the console output which contains two comments with the text `/* This should only exist once */` instead of one.

### 💻 Code
This is the same code as the repository that is linked above.

**Transformer that adds the synthetic comment (transform.ts):**
```ts
import * as ts from 'typescript';

const program = ts.createProgram(['./sample.ts'], {
  module: ts.ModuleKind.CommonJS,
  importHelpers: true,
  lib: ['dom', 'es2015'],
  target: ts.ScriptTarget.ES2017,
  declaration: true,
  experimentalDecorators: true,
  emitDecoratorMetadata: false,
});

const transformer: ts.TransformerFactory<ts.SourceFile> = context => {
  return sourceFile => {
    const visitor = (node: ts.Node): ts.Node => {
      if (ts.isClassDeclaration(node)) {
        const initializer = ts.createLiteral('hello');
        const prop = ts.createProperty(
            undefined, [ts.createToken(ts.SyntaxKind.StaticKeyword)],
            'customStaticProp', undefined, undefined, initializer);

        ts.setSyntheticLeadingComments(prop, [
          {
            kind: ts.SyntaxKind.MultiLineCommentTrivia,
            text: ' This should only exist once ',
            pos: -1,
            end: -1,
            hasTrailingNewLine: true,
          },
        ]);

        return ts.updateClassDeclaration(
            node, node.decorators, node.modifiers, node.name,
            node.typeParameters, node.heritageClauses, [...node.members, prop]);
      }

      return ts.visitEachChild(node, visitor, context);
    };

    return ts.visitNode(sourceFile, visitor);
  };
};

program.emit(
    undefined, ((fileName, outputText) => {
      console.log(outputText);
      console.log('---------------------', fileName);
    }),
    undefined, undefined, {before: [transformer]});
```

**File that is being transformed (sample.ts):**
```ts
function MyDecorator<T>(cls: T, ...args: any[]) {
  return cls;
}

@MyDecorator
export class Foo {
}
```

### 🙁 Actual behavior
TypeScript compiles the `customStaticProp` to the following:

```ts
/* This should only exist once */
/* This should only exist once */
Foo.customStaticProp = "hello";
```

### 🙂 Expected behavior
The `customStaticProp` should be compiled to:

```ts
/* This should only exist once */
Foo.customStaticProp = "hello";
```

### LINKED ISSUE #45817: Double emit of synthetic comments on class fields of decorated classes
# Bug Report

### 🔎 Search Terms

decorator transformer syntheticcomment emit field static

### 🕗 Version & Regression Information

- This changed between versions 4.3 and 4.4

### ⏯ Playground Link

no playground link, as this only occurs with synthetic comments (so needs a transformer)

### 💻 Code

As a unit test, in src/testRunner/unittests/transform.ts:

```ts
        testBaseline("transformSyntheticCommentOnStaticField", () => {
            return transpileModule(`
declare const Decorator: any;
@Decorator
class MyClass {
}
`, {
                transformers: {
                    before: [addStaticFieldWithComment],
                },
                compilerOptions: {
                    target: ScriptTarget.ES2015,
                    newLine: NewLineKind.CarriageReturnLineFeed,
                }
            }).outputText;

            function addStaticFieldWithComment(context: TransformationContext) {
                    return (sourceFile: SourceFile): SourceFile => {
                        return visitNode(sourceFile, rootTransform, isSourceFile);
                    };
                    function rootTransform<T extends Node>(node: T): Node {
                        if (isClassDeclaration(node)) {
                            const newMembers = [factory.createPropertyDeclaration(/* decorators */ undefined, [factory.createModifier(SyntaxKind.StaticKeyword)], "newField", /* questionOrExclamationToken */ undefined, /* type */ undefined, factory.createStringLiteral("x"))];
                            setSyntheticLeadingComments(newMembers[0], [{ kind: SyntaxKind.MultiLineCommentTrivia, text: "comment", pos: -1, end: -1, hasTrailingNewLine: true }]);
                            return factory.updateClassDeclaration(node, node.decorators, /* modifierFlags */ undefined, node.name, node.typeParameters, node.heritageClauses, newMembers);
                        }
                        return visitEachChild(node, rootTransform, context);
                    }
            }
        });

```

### 🙁 Actual behavior

```
  var MyClass = class MyClass {}
  /*comment*/
  /*comment*/
  MyClass.newField = "x";
  __decorate(MyClass, Decorator, ...)
```

### 🙂 Expected behavior

```
  var MyClass = class MyClass {}
  /*comment*/
  MyClass.newField = "x";
  __decorate(MyClass, Decorator, ...)
```

### PR DESCRIPTION
**Pull request by mprobst**
In TS4.4, when a transformer adds a static field with a synthetic
comment to a decorated class, TS prints the synthetic comment twice:

```
@Decorator
class MyClass {
  /* comment */ staticField = 'x';  // field and comment added by transformer
}
```

Becomes:

```
var MyClass = class MyClass {}
/*comment*/
/*comment*/
MyClass.newField = "x";
__decorate(MyClass, Decorator, ...)
```

This is because the classFields transformer calls `setOriginalNode(n,
propertyDeclaration)` on both the expression statement , and the
assignment expression contained in it, leading to the synthetic comment
appearing twice.

This change avoids the problem by only calling setOriginalNode on the
outer expression statement. All callers of the inner method take care to
`setOriginalNode` on the resulting container.

Fixes #45696.

### CODE REVIEW

**Comment by mprobst** [comment]
@rbuckton this might be related to recent changes made by you here?

Can you advise on the source map change? I'm not sure if that's OK.

**Comment by mprobst** [comment]
I believe the test suite failure is because I'm on a Mac, and the golden file update process produces `\r\n` in some locations here. Oddly enough the source map file in question is a mix of `\r\n` line endings and `\n` line endings, so it also isn't easy to fix using `dos2unix` or similar.

If we agree on the path, I can fix the file up manually or see if I can find a windows box to run this on.

**Comment by mprobst** [comment]
Friendly ping.

**Comment by mprobst** [comment]
Thanks for the feedback.

I've added the test you asked for, and made the change to explicitly delete synthetic comments (instead of dropping the `setOriginalNode`).

File: src/testRunner/unittests/transform.ts [hunk 1]
```diff
@@ -599,6 +599,63 @@ module MyModule {
         }, exports => {
             assert.equal(exports.stringLength, 5);
         });
+
+        function addStaticFieldWithComment(context: TransformationContext) {
+            return (sourceFile: SourceFile): SourceFile => {
+                return visitNode(sourceFile, rootTransform, isSourceFile);
+            };
+            function rootTransform<T extends Node>(node: T): Node {
+                if (isClassDeclaration(node) || isClassExpression(node)) {
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/transform.ts:608]:
```suggestion
                if (isClassLike(node)) {
```

**Pull request closed without merging**